idea/idea-fir/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtGenerateMembersHandler.kt

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.core.overrideImplement
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.PsiTreeUtil
+import org.jetbrains.kotlin.idea.core.insertMembersAfter
+import org.jetbrains.kotlin.idea.core.moveCaretIntoGeneratedElement
+import org.jetbrains.kotlin.idea.frontend.api.KtAnalysisSession
+import org.jetbrains.kotlin.idea.frontend.api.analyse
+import org.jetbrains.kotlin.idea.frontend.api.components.KtDeclarationRendererOptions
+import org.jetbrains.kotlin.idea.frontend.api.components.KtTypeRendererOptions
+import org.jetbrains.kotlin.idea.frontend.api.components.RendererModifier
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtCallableSymbol
+import org.jetbrains.kotlin.idea.frontend.api.tokens.HackToForceAllowRunningAnalyzeOnEDT
+import org.jetbrains.kotlin.idea.frontend.api.tokens.hackyAllowRunningOnEdt
+import org.jetbrains.kotlin.psi.KtCallableDeclaration
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.psiUtil.prevSiblingOfSameType
+import org.jetbrains.kotlin.utils.addToStdlib.firstIsInstanceOrNull
+
+abstract class KtGenerateMembersHandler : AbstractGenerateMembersHandler<KtClassMember>() {
+    @OptIn(HackToForceAllowRunningAnalyzeOnEDT::class)
+    override fun generateMembers(
+        editor: Editor,
+        classOrObject: KtClassOrObject,
+        selectedElements: Collection<KtClassMember>,
+        copyDoc: Boolean
+    ) {
+        // Using hackyAllowRunningOnEdt here because we don't want to pre-populate all possible textual overrides before user selection.
+        hackyAllowRunningOnEdt {
+            analyse(classOrObject) {
+                this.generateMembers(editor, classOrObject, selectedElements, copyDoc)
+            }
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun KtAnalysisSession.generateMembers(
+        editor: Editor,
+        currentClass: KtClassOrObject,
+        selectedElements: Collection<KtClassMember>,
+        copyDoc: Boolean
+    ) {
+        if (selectedElements.isEmpty()) return
+        val selectedMemberSymbols: Map<KtCallableSymbol, KtCallableDeclaration> = selectedElements.associate {
+            it.symbol to generateMember(currentClass.project, it, currentClass, copyDoc)
+        }
+
+        val classBody = currentClass.body
+        if (classBody == null) {
+            // TODO: implement proper referenceShortener
```

Comment:
We already have reference shortener in IDEA FIR, see `KtReferenceShortener`

Reply:
Ahh! I am lucky that I didn't decide to try implementing this!